core/src/main/scala/caliban/GraphQL.scala

```diff
@@ -95,68 +95,81 @@ trait GraphQL[-R] { self =>
-                                                                                request.operationName,
-                                                                                coercedVars,
-                                                                                config.skipValidation,
-                                                                                config.validations
-                                                                              )
-                                                              _            <- checkHttpMethod(config)(executionReq)
-                                                            } yield executionReq
-                  executionRequest                     <- wrap(validate)(validationWrappers, doc)
-                  op                                    = executionRequest.operationType match {
-                                                            case OperationType.Query        => schemaToExecute.query
-                                                            case OperationType.Mutation     => schemaToExecute.mutation.getOrElse(schemaToExecute.query)
-                                                            case OperationType.Subscription =>
-                                                              schemaToExecute.subscription.getOrElse(schemaToExecute.query)
-                                                          }
-                  execute                               = (req: ExecutionRequest) =>
-                                                            for {
-                                                              queryExecution <- Configurator.configuration.map(_.queryExecution)
-                                                              res            <- Executor.executeRequest(req, op.plan, fieldWrappers, queryExecution, features)
-                                                            } yield res
-                  result                               <- wrap(execute)(executionWrappers, executionRequest)
-                } yield result).catchAll(Executor.fail)
+                  doc          <- wrap(Parser.parseQuery)(parsingWrappers, request.query.getOrElse(""))
+                  coercedVars  <- coerceVariables(doc, request.variables.getOrElse(Map.empty))
+                  executionReq <- wrap(validation(request.operationName, coercedVars))(validationWrappers, doc)
+                  result       <- wrap(execution(schemaToExecute(doc), fieldWrappers))(executionWrappers, executionReq)
+                } yield result).catchAll(Executor.fail(_))
               )(overallWrappers, request)
           }
+
+        private def coerceVariables(doc: Document, variables: Map[String, InputValue])(implicit
+          trace: Trace
+        ): IO[ValidationError, Map[String, InputValue]] =
+          Configurator.configuration.flatMap { config =>
```

Comment:
The old code was reading the configuration only once for `skipValidation` and `enableIntrospection`. Not sure how much it matters.

Reply:
For `skipValidation`, it was actually reading it twice (once before `coerceVariables` and another before `Validator.prepare` which is the intended behaviour as Validation wrappers might change the `skipValidation` value.

This was one of the reasons I thought it'd be good to separate the phases into separate methods as well, because it felt like a bug waiting to happen if we accidentally read the state prior to the validation wrappers being applied